src/Http/Http.Abstractions/test/RouteHandlerInvocationContextOfTTests.cs

```diff
@@ -0,0 +1,60 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Http.Abstractions.Tests;
+
+public class RouteHandlerInvocationContextOfTTests
+{
+    [Fact]
+    public void ProhibitsActionsThatModifyListSize()
+    {
+        var context = new RouteHandlerInvocationContext<string, int, bool>(new DefaultHttpContext(), "This is a test", 42, false);
+        Assert.Throws<NotSupportedException>(() => context.Add("string"));
```

Comment:
Given this, I think we should change the type of `RouteHandlerInvocationContext.Arguments` to `IReadOnlyList<object?>`.

Reply:
I don't think so.

The type we use for `Arguments` will be used in both the generic and non-generic overloads. For the non-generic overload, we have to use `IList` to support in-place modification of items in the list.

One option is that we could change the behavior of the implementation so that if you're using the non-generic variant (more than 10 arguments) you're not able to make modifications to parameters. For scenarios where you're using ten arguments or less, you'll be able to modify in-place using the custom setter in the generic overloads.

Reply:
Actually, nevermind. We cannot do this anyway. The `IReadOnlyList` implementation does not support a setter on `this[index]` so we would be able to use it for generic variants anways.